XForm/XForm/Commands/CosmosUnescape.cs

```diff
@@ -0,0 +1,145 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.CodeAnalysis.Elfie.Model.Strings;
+using System;
+using System.Collections.Generic;
+using XForm.Data;
+using XForm.Extensions;
+using XForm.Query;
+
+namespace XForm.Commands
+{
+    internal class CosmosUnescapeCommandBuilder : IPipelineStageBuilder
+    {
+        public string Verb => "cosmosUnescape";
+        public string Usage => this.Verb;
+
+        public IDataBatchEnumerator Build(IDataBatchEnumerator source, WorkflowContext context)
+        {
+            if (context.Parser.HasAnotherPart)
+            {
+                throw new ArgumentOutOfRangeException($"{this.Verb} does not accept any arguments.");
+            }
+
+            return new CosmosUnescape(source);
+        }
+    }
+
+    public class CosmosUnescape : DataBatchEnumeratorWrapper
+    {
+        HashSet<int> _sourceColumns = new HashSet<int>();
+        Dictionary<String8, String8> _replacementMap;
+        String8 _bookmarkChar = String8.Convert("#", new byte[1]);
+
+        public CosmosUnescape(IDataBatchEnumerator source) : base(source)
+        {
+            for (int i = 0; i < source.Columns.Count; i++)
+            {
+                if (source.Columns[i].Type == typeof(String8))
+                {
+                    _sourceColumns.Add(i);
+                }
+            }
+        }
+
+        public Dictionary<String8, String8> ReplacementMap
+        {
+            get
+            {
+                if (_replacementMap == null)
+                {
+                    Dictionary<String8, String8> tempMap = new Dictionary<String8, String8>();
+                    tempMap.Add(String8.Convert("#R#", new byte[3]), String8.Convert("\r", new byte[1]));
+                    tempMap.Add(String8.Convert("#N#", new byte[3]), String8.Convert("\n", new byte[1]));
+                    tempMap.Add(String8.Convert("#TAB#", new byte[5]), String8.Convert("\t", new byte[1]));
+                    tempMap.Add(String8.Convert("#NULL#", new byte[6]), String8.Empty);
+                    tempMap.Add(String8.Convert("#HASH#", new byte[6]), String8.Convert("#", new byte[1]));
+
+                    _replacementMap = tempMap;
+                }
+
+                return _replacementMap;
+            }
+        }
+
+        public override Func<DataBatch> ColumnGetter(int columnIndex)
+        {
+            String8[] transformedArray = null;
+            String8Block block = new String8Block();
+
+            // Pass through columns other than the one being converted
+            if (!_sourceColumns.Contains(columnIndex)) return _source.ColumnGetter(columnIndex);
+
+            // Cache the function to get the source data
+            Func<DataBatch> sourceGetter = _source.ColumnGetter(columnIndex);
+
+            return () => Unescape(sourceGetter(), block, ref transformedArray);
```

Comment:
I think I'd move the String8Block.Clear out here to be a bit clearer - I like to see the thing which allocates the String8Block managing the clearing cycle. Not a big deal either way.